File: Data Structures & Algorithms/longest-consecutive-sequence/submission-4.java
class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for(int num : nums){
            set.add(num);
        }
        
        ArrayList<Integer> possibleStarts = new ArrayList<>();
        for(int num : nums){
            if(!set.contains(num-1)){
                possibleStarts.add(num);
            }
        }

        int maxLen = 0;
        for(int start : possibleStarts){
            int len = 1;
            while(set.contains(start+1)){
                start = start+1;
                len++;
            }
            maxLen = Math.max(maxLen, len);
        }
        return maxLen;
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [-1] * len(nums)

        def dfs(i):
            if i == len(nums):
                return 0
            if dp[i] != -1:
                return dp[i]
            res = 1
            for j in range(i+1, len(nums)):
                if nums[j] > nums[i]:
                    res = max(res, 1+dfs(j))
            dp[i] = res
            return res

        res = 0
        for i in range(len(nums)):
            res = max(res, dfs(i))
        
        return res